File: 0012 整数转罗马数字/main.py
class Solution:
    def intToRoman(self, num: int) -> str:
        Romo = [("M", 1000), ("D", 500), ("C", 100),
                ("L", 50), ("X", 10), ("V", 5), ("I", 1)]
        s = ""
        for i in range(len(Romo)):
            times = num // Romo[i][1]
            print(Romo[i], times,num,s)
            s += times*Romo[i][0]
            num %= Romo[i][1]
            if(Romo[i][1] != 1):
                if(str(Romo[i][1])[0] == "1" and num//(Romo[i][1]*9//10) != 0):
                    print("2")
                    num -= Romo[i][1]*9//10
                    s += Romo[i+2][0] + Romo[i][0]
                elif(str(Romo[i][1])[0] == "5" and num//(Romo[i][1]*4//5) != 0):
                    print("1")
                    num -= Romo[i][1]*4//5
                    s += Romo[i+1][0] + Romo[i][0]
        return s


a = Solution()
print(a.intToRoman(58))

File: 0012 整数转罗马数字/main.cpp
class Solution {
public:
    string intToRoman(int num) {
        vector<int> numtable{1000,900,500,400,100,90,50,40,10,9,5,4,1};
        vector<string> strtable{"M","CM","D","CD","C","XC","L","XL","X","IX","V","IV","I"};
        string res="";
        int n=0;
        for(int i =0;i<numtable.size();i++){
            n=num / numtable[i];
            if( n != 0 ){
              for(int j=0;j<n;j++){
                res.append(strtable[i]);
              }
              num=num%numtable[i];
            }
        }
        return res;
    }
};